### PR TITLE
chore(dockerfile): make docker build multistage

### PR DESCRIPTION
**Pull request by hazzadous**
The built image is >4GB uncompressed atm, I'm pretty sure there is a lot
of cruft.

Plan is to split out the django, frontend, plugin-server builds and
hopefully get some gains in there to not include build deps.

This gets the uncompressed image down to 1.48GB, compressed around 500MB 🤞 

It also adds caching to the test image build workflow.

Some important changes there are slightly tangential:

 1. we run on master branch, to ensure we have the master image for caching
 2. we run the E2E tests in the same workflow, as using `workflow_run` trigger events seemed to be more complicated than expected
 3. to ensure that the image is functional, we run the E2E tests with it

NOTE: I've also kicked off a test with the helm chart [here](https://github.com/PostHog/charts-clickhouse/pull/449/checks)
NOTE 2: if caching works, the total runtime of both the build and E2E is ~10mins. If it doesn't though, it's 30 mins. E2E previously I think is around 20 mins currently.

TODO:

 * distribute staticfiles only by CDN(? this might not work for super security conscious installs)
 * distribute a separate plugin server image
 * we are doubling up on staticfiles with the collect step, we can probably get rid of the originals somehow but I'm not sure exactly how
 * clean up the user handling, we switch back an forth but could probably do better
 * use this image for tests, to ensure the image build is correct
 * caching is good sometimes, but not if you want security patches to be applied. Make sure prod has security patches applied frequently. I've actually only enabled caching on branches so that's not an issue atm.
 * remove old image tags so orphan layers can be pruned
 * We also build the image elsewhere for releases, we can just reuse this build instead

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: .github/workflows/docker-image-test.yml [hunk 1]
```diff
@@ -16,19 +16,57 @@ jobs:
...
+                  password: ${{ secrets.GITHUB_TOKEN }}
+
             - name: Build
               id: docker_build
               uses: docker/build-push-action@v2
               with:
-                  push: false
-                  load: true
-                  tags: posthog/posthog:testing
+                  push: true
```
Comment (pauldambra) [inline .github/workflows/docker-image-test.yml:61]: `load:true` was needed so the output image size job worked

_but_ at a quarter of the size maybe we just delete the output image size job

Comment (hazzadous) [reply .github/workflows/docker-image-test.yml:61]: ahhhh ok I'll revert

**Review by fuziontech — APPROVED** [review]
This looks really exciting to me. Let's land this on Monday and get this in! 🚢 it

File: production.Dockerfile [hunk 2]
```diff
@@ -15,25 +65,15 @@ WORKDIR /code
...
-    && npm install -g yarn@1
+    "xmlsec~=1.2"
 
-# Install SAML dependencies
-RUN apk --update --no-cache add \
-    "libxml2-dev~=2.9" \
-    "xmlsec~=1.2" \
-    "xmlsec-dev~=1.2"
+# Install SAML runtime dependencies
+RUN apk --update --no-cache add "xmlsec~=1.2"
```
Comment (hazzadous) [inline production.Dockerfile:76]:
```suggestion
```

File: production.Dockerfile [hunk 3]
```diff
@@ -3,6 +3,57 @@
...
+
+#
+# Build the plugin-server artifacts. Note that we still need to install the
+# runtime deps in the main image
+#
+FROM node:16.15-alpine3.14 AS plugin-server
+
+WORKDIR /code/plugin-server
+
+# Install python, make and gcc as they are needed for the yarn install
```
Comment (guidoiaquinti) [inline production.Dockerfile:31]: Nit: also Python?

**Comment by guidoiaquinti** [comment]
Still quite a mess but way better than before. Nice work (as usual) @hazzadous !

**Review by guidoiaquinti — APPROVED** [review]

File: production.Dockerfile (hunk 3 above)
Comment (hazzadous) [reply production.Dockerfile:31]: As far as I could tell yes python is also needed. I could be wrong there though.

**Review by guidoiaquinti — APPROVED** [review]

File: production.Dockerfile [hunk 4]
```diff
@@ -46,70 +85,73 @@ RUN apk --update --no-cache add \
...
-# Add a dedicated 'posthog' user and group, move files into its home dir and set the
-# proper file permissions. This alleviates compliance issue for not running a
-# container as 'root'
-RUN addgroup -S posthog && \
-    adduser -S posthog -G posthog && \
-    mv /code /home/posthog && \
-    chown -R posthog:1000 /home/posthog/code
-WORKDIR /home/posthog/code
+# Add in the compiled plugin-server
+COPY --from=plugin-server /code/plugin-server/dist/ ./dist/
```
Comment (benjackwhite) [inline production.Dockerfile:138]: Is it really the case that the plugin-server does not need `node_modules` ? I understand it for the frontend as that is all bundled code but surely nodejs uses it in the plugin server?

Comment (hazzadous) [reply production.Dockerfile:138]: I suspect it does need them, so we install them in the final image as well.

Note that we now also use the image in the E2E tests, which should catch this(?)

Comment (hazzadous) [reply production.Dockerfile:138]: ( see https://github.com/PostHog/posthog/pull/10488/files#diff-55d06c59f3552514747959fc163ad2940d33f0757fa48cf363580af437dee6a3R131 )

**Pull request merged**